### DISCUSSION TITLE
How in the world can I get an header?

### DISCUSSION THREAD

**Discussion by wavz217** [rwf2/Rocket#2551] [Questions] [answered]
I tried everything, but I'm too dumb to understand how request guards work.
I really need to get the X-Real-IP header as String.

Yes, I have seen documentation and other GitHub Discussion posts. I still can't understand a bit.

Please help me out!

**Comment by robertwayne** [comment, marked as answer]
Rocket provides a built-in guard for getting the IP Address, see [here](https://api.rocket.rs/v0.5-rc/rocket/request/trait.FromRequest.html) and [here](https://api.rocket.rs/v0.5-rc/rocket/request/struct.Request.html#method.client_ip). This - under the hood - defaults to extracting the  "X-Real-IP" header.

```rust
#[get("/")]
async fn get_ip(ip: std::net::IpAddr) -> String {
    format!("Your IP is: {}", ip)
}

#[launch]
fn rocket() -> _ {
    rocket::build().mount("/", routes![get_ip])
}
```

However, if you wish to manually extract a header, you have access to `.headers()` in a request guard (see [here](https://api.rocket.rs/v0.5-rc/rocket/request/struct.Request.html#method.headers)). I feel like this [example](https://api.rocket.rs/v0.5-rc/rocket/request/trait.FromRequest.html#example-1) is quite self-explanatory, and could be modified to your use-case quite easily by changing just the header names. If you have questions on that specifically, feel free to ask.
